### ISSUE TITLE
Cannot find type definition file for './webdriverio-core

### ISSUE THREAD

**Issue by gitowiec** [webdriverio/webdriverio#3639]
**Environment (please complete the following information):**
 - **WebdriverIO version:** 5.6.3

**Describe the bug**
I installed latest webdriver and I am trying to compile it, TypeScript compiler rises error.
Cannot find type definition file for './webdriverio-core

**Comment by christian-bromann** [comment]
> I installed latest webdriver

`webdriver` doesn't use the webdriverio-core definition, did you install the `webdriverio` package? Can also make sure to update all packages?

**Comment by gitowiec** [comment]
Here is bugs I got when installed 5.6.2

![webstorm](https://user-images.githubusercontent.com/1553195/53497940-96786c00-3aa5-11e9-99f4-c70d9104d7c2.png)

**Comment by gitowiec** [comment]
Here are devDependencies I just installed

    "@wdio/cli": "5.6.2",
    "@wdio/dot-reporter": "5.6.0",
    "@wdio/local-runner": "5.6.2",
    "@wdio/mocha-framework": "5.6.0",
    "@wdio/selenium-standalone-service": "5.6.0",
    "@wdio/spec-reporter": "5.6.0",
    "@wdio/sync": "5.6.0",
    "webdriverio": "5.6.2",

**Comment by gitowiec** [comment]
What TypeScript version should I use with 5.6.3 or 5.6.x in general? There is missing "typeScriptVersion": "..."  field in package.json :(

**Comment by christian-bromann** [comment]
@mgrybyk can you comment on that ^^^ ?

@gitowiec you always need to make sure to grab all patch updates, so modify your dev dependencies to:
```json
"@wdio/cli": "~5.6.2",
"@wdio/dot-reporter": "~5.6.0",
"@wdio/local-runner": "~5.6.2",
"@wdio/mocha-framework": "~5.6.0",
"@wdio/selenium-standalone-service": "~5.6.0",
"@wdio/spec-reporter": "~5.6.0",
"@wdio/sync": "~5.6.0",
"webdriverio": "~5.6.2",
```

**Comment by gitowiec** [comment]
We don't use semver prefixes, we lerned that they are unreliable (but it's just not important here).
I just updated everything like this:
    "@wdio/cli": "5.6.3",
    "@wdio/dot-reporter": "5.6.0",
    "@wdio/local-runner": "5.6.3",
    "@wdio/mocha-framework": "5.6.3",
    "@wdio/selenium-standalone-service": "5.6.3",
    "@wdio/spec-reporter": "5.6.0",
    "@wdio/sync": "5.6.3",
    "webdriverio": "5.6.3",

and added latest TypeScript.
And I see a lot of change in API.
`WebdriverIO.Element` is no more generic, so I have to remove a lot of `<void>` declarations.
And $ function does not returns Element any more, it started to return `Promise<WebdriverIOAsync.Element>` - so how should I use it? With `async await`?

**Comment by gitowiec** [comment]
And this part of changelog.md
![changelog](https://user-images.githubusercontent.com/1553195/53502282-0854b380-3aae-11e9-8082-f982b83f69a1.png)
Should be marked with fire icon, because here https://github.com/webdriverio/webdriverio/pull/3577 it says "Please note that it is a braking change for existing TypeScript users that work in sync mode" !! :(
This changes are really tough, so, You just understand me that one can't relay on semver (it's just a phantasy)

**Comment by CrispusDH** [comment]
If you use `@wdio/sync` => 

![image](https://user-images.githubusercontent.com/12623263/53502924-be70cb00-3ab7-11e9-9511-080024c8d46e.png)

**Comment by gitowiec** [comment]
Yep I figured it out, but thanks!

**Comment by gitowiec** [comment]
Should I uninstall webdriverio package? I get TypeScript conflict with `TS2300: Duplicate identifier 'BrowserObject'`

![conflict](https://user-images.githubusercontent.com/1553195/53504365-1c021900-3ab2-11e9-931f-d31812b107bc.png)

**Comment by mgrybyk** [comment]
@gitowiec just remove webdriverio from tsconfig typings and remove all explicit imports in code

**Comment by gitowiec** [comment]
I did it.
Here is my `tsconfig.json` (we use monorepo)

<pre>
{
  "compilerOptions": {
    "outDir": "./dist/",
    "sourceMap": true,
    "noImplicitAny": false,
    "strictNullChecks": true,
    "experimentalDecorators": true,
    "module": "commonjs",
    "target": "es5",
    "jsx": "react",
    "baseUrl": ".",
    "types": [
      "node",
      "mocha",
      "@wdio/sync"
    ],
    "lib": [
      "dom",
      "esnext"
    ],
    "allowJs": true
  },
  "include": [
    "node_modules/awesome-typescript-loader/lib/runtime.d.ts",
    "./../../node_modules/components/src/**/*",
    "./../../node_modules/cash-report-template/src/**/*",
    "./../../node_modules/invoice-template/src/**/*",
    "./src/config.d.ts",
    "./src/**/*",
    "../../node_modules/transport"
  ],
  "exclude": [
    "node_modules",
    "./../../node_modules",
    "dist",
    "./../../node_modules/cash-report-template/node_modules/styled-components",
    "./../../node_modules/invoice-template/node_modules/styled-components",
    "e2e"
  ]
}
</pre>

**Comment by mgrybyk** [comment]
So, what problems so you have now?
Intelliji works bad with typescript.

**Comment by gitowiec** [comment]
Ok, I overlooked two imports `import * as WebdriverIO from 'webdriverio';` I corrected them to `import * as WebdriverIO from '@wdio/sync';` and now it looks like there are no errors releated to WebdriverIO.
Thank You very much for your time and patience.
Where I can check in the docs what is the minimum TypeScript version required to use "@wdio/sync": "5.6.3" typings?

**Comment by mgrybyk** [comment]
I'm using latest, not sure to be honest.
It have to be one of recent versions, we are using lots of its features

Btw, why do you importing webdriverio?
@gitowiec

**Comment by mgrybyk** [comment]
Such import should not work `import * as WebdriverIO from '@wdio/sync'; `

You'll import @wdio/sync with webdriverio typings, doesn't make sense to me.

I think we have some confusion here.

**Comment by gitowiec** [comment]
I have imported that namespace to add type here

```
import { getTestSelector } from '../../../../e2e/src/helpers/selectors';
import {
  PAGE_COMPANY_TAB_LABEL_PAYMENTS_TEST_ID,
  PAGE_COMPANY_TABS_LABELS_CONTAINER_TEST_ID,
} from '../../../modules/Companies/pages/PageCompany.test.consts';
import * as WebdriverIO from '@wdio/sync';

export class PageTabsPageObject {

  private paymentsTab: WebdriverIO.Element;

  getTabsContainer() {
    return $(getTestSelector(PAGE_COMPANY_TABS_LABELS_CONTAINER_TEST_ID));
  }

  getPaymentTab() {
    const paymentTab = this.getTabsContainer().$(getTestSelector(PAGE_COMPANY_TAB_LABEL_PAYMENTS_TEST_ID));
    paymentTab.waitForExist(10_000, false, 'Brak zakładki Płatności');
    this.paymentsTab = paymentTab;
  }

  goToPayments() {
    this.getPaymentTab();
    this.paymentsTab.click();
  }

}
```

**Comment by mgrybyk** [comment]
You don't beed to import webdriverio to do that! It will cause conflicts

**Comment by christian-bromann** [comment]
@gitowiec do you still see any errors or can this be closed?

**Comment by GenadyMgr** [comment]
@christian-bromann I'm having similar problem, I'm using @wdio/sync, and I added "@wdio/sync" to the types in tsconfig.js
But I still cannot find types, I'm getting the following errors:
```
[11:37:45 PM] Starting compilation in watch mode...

node_modules/@wdio/sync/webdriverio.d.ts:1:23 - error TS2688: Cannot find type definition file for './webdriverio-core'.

1 /// <reference types="./webdriverio-core"/>
                        ~~~~~~~~~~~~~~~~~~

node_modules/@wdio/sync/webdriverio.d.ts:3:22 - error TS2503: Cannot find namespace 'WebDriver'.

3 type BrowserObject = WebDriver.ClientOptions & WebDriver.Client & WebdriverIO.Browser;
                       ~~~~~~~~~

node_modules/@wdio/sync/webdriverio.d.ts:3:48 - error TS2503: Cannot find namespace 'WebDriver'.

3 type BrowserObject = WebDriver.ClientOptions & WebDriver.Client & WebdriverIO.Browser;
                                                 ~~~~~~~~~

node_modules/@wdio/sync/webdriverio.d.ts:3:67 - error TS2314: Generic type 'Browser<T>' requires 1 type argument(s).

3 type BrowserObject = WebDriver.ClientOptions & WebDriver.Client & WebdriverIO.Browser;
                                                                    ~~~~~~~~~~~~~~~~~~~

node_modules/@wdio/sync/webdriverio.d.ts:7:19 - error TS2503: Cannot find namespace 'WebDriver'.

7         options?: WebDriver.Options,
                    ~~~~~~~~~

node_modules/@wdio/sync/webdriverio.d.ts:12:30 - error TS2694: Namespace 'WebdriverIO' has no exported member 'MultiRemoteOptions'.

12         options: WebdriverIO.MultiRemoteOptions
                                ~~~~~~~~~~~~~~~~~~

node_modules/@wdio/sync/webdriverio.d.ts:13:8 - error TS2503: Cannot find namespace 'WebDriver'.

13     ): WebDriver.Client;
          ~~~~~~~~~

node_modules/@wdio/sync/webdriverio.d.ts:17:50 - error TS2314: Generic type 'Element<T>' requires 1 type argument(s).

17 declare function $(selector: string | Function): WebdriverIO.Element;
                                                    ~~~~~~~~~~~~~~~~~~~

node_modules/@wdio/sync/webdriverio.d.ts:18:51 - error TS2314: Generic type 'Element<T>' requires 1 type argument(s).

18 declare function $$(selector: string | Function): WebdriverIO.Element[];
                                                     ~~~~~~~~~~~~~~~~~~~

src/helpers/utils.ts:183:54 - error TS2339: Property 'Reuters' does not exist on type 'typeof Providers'.

183         return [Providers.Celebrity_Media, Providers.Reuters, Providers.Bloomberg,
                                                         ~~~~~~~

src/helpers/utils.ts:184:34 - error TS2339: Property 'Giphy' does not exist on type 'typeof Providers'.

184         Providers.AFP, Providers.Giphy, Providers.GMG, Providers.AP, Providers.USAToday, Providers.Gannett, Providers.Abaca];
                                     ~~~~~

src/specs/api/settings-spec.ts:36:72 - error TS2345: Argument of type '{ Accept: string; }' is not assignable to parameter of type 'string | FormData'.
  Type '{ Accept: string; }' is not assignable to type 'FormData'.
    Property 'append' is missing in type '{ Accept: string; }'.

36             yield api.sendInternalApiRequest(url, settingParam, "GET", headers);
                                                                          ~~~~~~~

```

**Comment by mgrybyk** [comment]
@GenadyMgr what version of typescript to you have?

**Comment by mgrybyk** [comment]
@GenadyMgr @gitowiec 
minimal version is 3.2.1

**Comment by mgrybyk** [comment]
https://github.com/webdriverio/webdriverio/pull/3643

**Comment by gitowiec** [comment]
Now I have TypeScript version set to 3.3333

**Comment by GenadyMgr** [comment]
@mgrybyk 
After updating to the latest typescript the issue was fixed. thanks!

**Comment by mgrybyk** [comment]
> Now I have TypeScript version set to 3.3333
You mean 3.3.3333 ? :)
Great! Is you problem resolved @gitowiec ?

**Comment by christian-bromann** [comment]
Closing then.

**Comment by FelixZilber** [comment]
@christian-bromann I have updated the dependencies. But still have this problem. In my case I have a util repo that wraps wdio functionality. When compiling TS it generates d.ts file. this contains
 /// <reference types="./webdriverio-core" />
any idea what it can be?

**Comment by mgrybyk** [comment]
Please provide us with reproducible example @FelixZilber 

Common mistakes:
- explicit imports of webdriverio
- wrong TypeScript version
- wrong types add in tsconfig json
- some of wdio packages are not updated, it is required to update @wdio/sync (if used) @wdio/webdriver and webdriverio to the latest versions. Sometimes old versions persist in package lock json

Minimal reproducible example will help a lot

**Comment by CrispusDH** [comment]
@mgrybyk , it might be useful to add "Common mistakes" to the documentation about TypeScript integration. What do you think?

**Comment by mgrybyk** [comment]
Maybe, let's get more statistics first :)

**Comment by FelixZilber** [comment]
@mgrybyk this is the repo I get the failure in : https://github.com/cloudinary/wdio-allure-ts/tree/CLD-10690-upgrade-to-v5
the d.ts file : https://github.com/cloudinary/wdio-allure-ts/blob/CLD-10690-upgrade-to-v5/lib/src/commons/BrowserUtils.d.ts
there is no import of webdriverIO, but there is a @wdio/sync import ...

While this repo compiles just fine, importing BrowserUtils in other repo will cause :
`node_modules/wdio-allure-ts/lib/src/commons/BrowserUtils.d.ts:1:23 - error TS2688: Cannot find type definition file for './webdriverio-core'.

1 /// <reference types="./webdriverio-core" />`

**Comment by mgrybyk** [comment]
Taking a look

**Comment by mgrybyk** [comment]
@FelixZilber I don't have any problems in your project.
`npm run ts_compile` passes.

Most likely you have some problems with your node_modules folder.
Try deleting it and re-installing modules `npm i`

**Comment by FelixZilber** [comment]
@mgrybyk that's exactly my problem. this project compilation passes, but while importing it in other repo it fails on compilation.
You can see the error here: 
[https://travis-ci.org/cloudinary/wdio-allure-ts-example/builds/501992206?utm_source=github_status&utm_medium=notification](url)

Branch: [https://github.com/cloudinary/wdio-allure-ts-example/tree/wdio-v5](url)

**Comment by mgrybyk** [comment]
@FelixZilber to be honest I don't understand your problem.
You have some issues with your configuration, there is nothing to do with webdriverio.

According to log there are problems with BrowserUtils.d.ts

Fix errors there first

**Comment by mgrybyk** [comment]
Remove reference to 

/// <reference types="./webdriverio-core" />

Obviously it won't work

**Comment by FelixZilber** [comment]
@mgrybyk the reference added on TS compilation. Thanks anyway. Will update with solution

**Comment by mgrybyk** [comment]
Thanks, it would be really helpful.
I have no idea why reference is added on complication, never seen such behavior

**Comment by FelixZilber** [comment]
@mgrybyk Looks like I found the reason for my failure but did not find the solution yet.
As I showed in my BrowserUtils.d.ts there is a reference to webdriverio-core. this added since BrowserUtils.ts import from @wdio/sync. And compilation passed just fine. (Let's call it SubProject)

The problem happen when I import BrowserUtils in other project (lets call it ParentProject) that also have @wdio/sync dependency.
In such case npm resolution steps in. It recognize same dependency and keeps it only in ParentProject. Since the reference is relative, it does not exist under SubProject anymore ==> compilation error.

Does it make sense ?

**Comment by mgrybyk** [comment]
@FelixZilber not really to be honest.
You should bot refer to webdriverio-core.
This is internal file.

**Comment by FelixZilber** [comment]
@mgrybyk I have 2 workarounds that work:
1) adding reference types="@wdio/sync/webdriverio-core" to BrowserUtils.ts which does not sound right(but works) 
2) use absolute reference type to reference types="@wdio/sync/webdriverio-core"
in webdriverio.d.ts but this one part of the @wdio/sync

**Comment by mgrybyk** [comment]
Both are bad @FelixZilber

Import module properly as @wdio/sync

Do not refer to files inside package, it may change at any time

**Comment by FelixZilber** [comment]
@mgrybyk I agree that the first solution is bad.
I did import @wdio/sync, but I got relative path to the reference.

Shouldn't the reference in webdriverio.d.ts be absolute ?
https://github.com/webdriverio/webdriverio/blob/11da9ad430f42d8d03ace1d9f7b83ac29a8ce7f8/packages/wdio-sync/webdriverio.d.ts#L1

**Comment by mgrybyk** [comment]
How can it be absolute?

**Comment by mgrybyk** [comment]
If you import @wdio/sync as module you shouldn't care about any internal paths

**Comment by FelixZilber** [comment]
> How can it be absolute? Not sure if it is possible

`/// <reference types="@wdio/sync/webdriverio-core"/> `

**Comment by mgrybyk** [comment]
So why can't you just import
```/// <reference types="@wdio/sync"/>```

?

**Comment by FelixZilber** [comment]
because it will have next result in my d.ts :
`/// <reference types="./webdriverio-core" />
`

which is wrong 😩

**Comment by mgrybyk** [comment]
Oh man

**Comment by mgrybyk** [comment]
So how can we handle it?

**Comment by FelixZilber** [comment]
https://github.com/webdriverio/webdriverio/pull/3688

How about that?

**Comment by mgrybyk** [comment]
Looks good, let's proceed conversation there.
Thanks

**Comment by khyati2010** [comment]
While building my project , I am getting 

`ndi-mb-041:e2e khyati.sehgal$ yarn build
yarn run v1.22.10
warning ../../../../package.json: No license field
$ npm run lint && rimraf lib && npm run moveJsonFiles && tsc --outDir lib
npm WARN lifecycle The node binary used for scripts is /var/folders/0z/13848h057r56wcwp808shyhr3w4d0j/T/yarn--1626420714236-0.4627629329507601/node but npm is using /usr/local/bin/node itself. Use the `--scripts-prepend-node-path` option to include the path for the node binary npm was executed with.

  > QA-automation@0.1.0 lint /Users/khyati.sehgal/Documents/workspace/web-app/e2e
> eslint --fix src/**/*.ts

npm WARN lifecycle The node binary used for scripts is /var/folders/0z/13848h057r56wcwp808shyhr3w4d0j/T/yarn--1626420714236-0.4627629329507601/node but npm is using /usr/local/bin/node itself. Use the `--scripts-prepend-node-path` option to include the path for the node binary npm was executed with.

> semo-automation@0.1.0 moveJsonFiles /Users/khyati.sehgal/Documents/workspace/web-app/e2e
> copyfiles -u 1 src/selectors/* src/config/* lib

error TS2688: Cannot find type definition file for './wdio'.


Found 1 error.

error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.`

I dint upgraded anything, neither I changed my config or package file. I cleared cache, removed lock files , with no luck. any suggestions for this here?
